pkg/reconciler/kubequota/kubequota_controller.go

```diff
@@ -0,0 +1,423 @@
+	quotaRecalculationPeriod time.Duration
+	// fullResyncPeriod controls how often the dynamic informers do a full resync
+	fullResyncPeriod time.Duration
+
+	workersPerLogicalCluster int
+
+	// lock guards the fields in this group
+	lock                       sync.RWMutex
+	cancelFuncs                map[logicalcluster.Name]func()
+	resourceQuotaEventHandlers map[logicalcluster.Name]cache.ResourceEventHandler
+
+	delegatingEventHandler *delegatingEventHandler
+
+	discoveryStopsLock sync.Mutex
+	discoveryStops     map[logicalcluster.Name]chan struct{}
+
+	// For better testability
+	getClusterWorkspace func(key string) (*tenancyv1alpha1.ClusterWorkspace, error)
+}
+
+// NewController creates a new Controller.
+func NewController(
+	clusterWorkspacesInformer tenancyinformers.ClusterWorkspaceInformer,
+	kubeClusterClient kubernetes.ClusterInterface,
+	resourceQuotaInformer coreinformers.ResourceQuotaInformer,
+	informerFactoryForQuota informerfactory.InformerFactory,
+	quotaRecalculationPeriod time.Duration,
+	fullResyncPeriod time.Duration,
+	workersPerLogicalCluster int,
+	informersStarted <-chan struct{},
+) (*Controller, error) {
+	c := &Controller{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+
+		resourceQuotaInformer:   resourceQuotaInformer,
+		informerFactoryForQuota: informerFactoryForQuota,
+		kubeClusterClient:       kubeClusterClient,
+		informersStarted:        informersStarted,
+
+		quotaRecalculationPeriod: quotaRecalculationPeriod,
+		fullResyncPeriod:         fullResyncPeriod,
+
+		workersPerLogicalCluster: workersPerLogicalCluster,
+
+		cancelFuncs:                map[logicalcluster.Name]func(){},
+		resourceQuotaEventHandlers: map[logicalcluster.Name]cache.ResourceEventHandler{},
+		delegatingEventHandler: &delegatingEventHandler{
+			eventHandlers: map[schema.GroupResource]map[logicalcluster.Name]cache.ResourceEventHandler{},
+		},
+
+		discoveryStops: map[logicalcluster.Name]chan struct{}{},
+
+		getClusterWorkspace: func(key string) (*tenancyv1alpha1.ClusterWorkspace, error) {
+			return clusterWorkspacesInformer.Lister().Get(key)
+		},
+	}
+
+	clusterWorkspacesInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.enqueue(obj)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				c.enqueue(oldObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.enqueue(obj)
+			},
+		},
+	)
+
+	// This sets up a single global event handler on ResourceQuotas. When a resource quota controller is created for
+	// a logical cluster, it registers its own event handler, which is stored in Controller.resourceQuotaEventHandlers.
+	// When an event comes in, this handler looks up the actual handler for the object's logical cluster and invokes it.
+	resourceQuotaInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
```

Comment:
This looks a lot like `delegatingEventHandler` - what's different?

Reply:
We have generics soon ;)

Reply:
Got rid of this.